internal/infrastructure/kubernetes/applier/apply.go

```diff
@@ -122,7 +124,7 @@ func (i *Instance) CreateOrUpdateService(ctx context.Context, svc *corev1.Servic
 		}
 	} else {
 		// Update if current value is different.
-		if !reflect.DeepEqual(svc.Spec, current.Spec) {
+		if !utils.CompareSvc(svc, current) {
```

Comment:
there are two bugs associated with this logic
1. we are comparing the entire `Spec` here, instead  we should be comparing _all_ the fields the infra layer sets or skip the fields the infra layer does not set https://pkg.go.dev/k8s.io/api/core/v1#ServiceSpec.

2. when updating the service, we need to set immutable fields like `ClusterIP` and `ClusterIPs`
```
// clusterIP is the IP address of the service and is usually assigned
	// randomly. If an address is specified manually, is in-range (as per
	// system configuration), and is not in use, it will be allocated to the
	// service; otherwise creation of the service will fail. This field may not
	// be changed through updates unless the type field is also being changed
	// to ExternalName (which requires this field to be blank) or the type
	// field is being changed from ExternalName (in which case this field may
	// optionally be specified, as describe above).  Valid values are "None",
	// empty string (""), or a valid IP address. Setting this to "None" makes a
	// "headless service" (no virtual IP), which is useful when direct endpoint
	// connections are preferred and proxying is not required.  Only applies to
	// types ClusterIP, NodePort, and LoadBalancer. If this field is specified
	// when creating a Service of type ExternalName, creation will fail. This
	// field will be wiped when updating a Service to type ExternalName.
	// More info: https://kubernetes.io/docs/concepts/services-networking/service/#virtual-ips-and-service-proxies
	// +optional
	ClusterIP [string](https://pkg.go.dev/builtin#string) `json:"clusterIP,omitempty" protobuf:"bytes,3,opt,name=clusterIP"`
```

Reply:
1, ok, i'll add imutable fields compare we set. 
2, when updating the svc, is these be better that just patch what we need rather than set everything to the new svc?
3, actually，when we use loadbalacer by default, svc.Spec.ports[*].nodePort will be set an random number by k8s, so it's not just about user modified the svc type.

Reply:
Reg 2. `Patch` is fine, my only request is to keep the code simple so it can be enhanced with more fields in the future
Reg 3. its okay to undo what user did, to solve this, we must
* support `Nodeport` natively
* support BYO Envoy Svc and Envoy Deployment and disable managed infra

Reply:
raise another pr for support `NodePort` natively might be a better way?

Reply:
sure please go ahead @spwangxp